File: static_media/code/a-question-of-ingestion.cpp
#include <functional>
#include <iostream>
#include <unordered_map>
#include <vector>

using ll = long long;

// dp[c][k] = the highest total calories we can consume with a capacity
ll dp[101][101];

int main() {
  ll n, m;
  std::cin >> n >> m;
  std::vector<ll> courses(n);
  for (auto &c : courses)
    std::cin >> c;

  std::unordered_map<ll, ll> cap_idx;

  for (ll amt = m, ate = 0; amt > 0; ++ate, amt = 2 * amt / 3)
    cap_idx[amt] = ate;

  for (int r = 0; r < cap_idx.size(); ++r)
    for (int c = 0; c < n; ++c)
      dp[r][c] = -1;

  std::function<ll(ll, ll, int)> select = [&](ll cap, ll prev_cap, int course) {
    if (cap == 0 || course >= n)
      return 0LL;
    ll &best_option = dp[cap_idx[cap]][course];
    if (best_option != -1)
      return best_option;
    ll skip1 = select(prev_cap, cap, course + 1);
    ll skip2 = select(m, cap, course + 2);
    ll take =
        std::min(courses[course], cap) + select(2 * cap / 3, cap, course + 1);
    best_option = std::max({skip1, skip2, take});
    return best_option;
  };

  ll ans = select(m, m, 0);

  std::cout << ans << '\n';
  return 0;
}

File: static_media/code/a-question-of-ingestion.py
dp = [[-1 for i in range(101)] for j in range(101)]

# n,     m
courses, max_hunger = map(int, input().split())
cals_per_course = list(map(int, input().split()))

max_consecutive_courses = 0
h = max_hunger
while (h > 0):
    h = h*2//3
    max_consecutive_courses += 1

# how much can we eat after `k` consecutive courses?
def hunger(k):
    appetite = max_hunger
    for _ in range(k):
        appetite = 2*appetite//3
    return appetite

# while `score` is recursive, recursion depth will never exceed 1 due to our call order
def score(course, consecutive_courses):
    if (course >= courses or k >= max_consecutive_courses):
        return 0
    if (dp[course][consecutive_courses] != -1):
        return dp[course][k]
    return score(c+1, max_hunger+1) + min(hunger(consecutive_courses), cals_per_course[course])

for c in range(courses-1, -1, -1):
    for k in range(max_consecutive_courses-1, -1, -1):
        take  = score(course=c,   consecutive_courses=k)
        skip1 = score(course=c+1, consecutive_courses=max(0,k-1))
        skip2 = score(course=c+2, consecutive_courses=0)
        dp[c][max_hunger] = max(take, skip1, skip2)

print(dp[0][0])
